janusgraph-core/src/main/java/org/janusgraph/graphdb/query/graph/GraphCentricQueryBuilder.java

```diff
@@ -301,7 +301,7 @@ public GraphCentricQuery constructQueryWithoutProfile(final ElementCategory resu
             indexLimit = Math.min(hardMaxLimit,
                 QueryUtil.adjustLimitForTxModifications(tx, conditions.numChildren() - coveredClauses.size(), indexLimit));
             query = new BackendQueryHolder<>(selectedIndex.getQuery().updateLimit(indexLimit),
-                    coveredClauses.size() == conditions.numChildren(), selectedIndex.isSorted());
+                    coveredClauses.containsAll(conditions) || coveredClauses.contains(conditions), selectedIndex.isSorted());
```

Comment:
when `conditions` is an `Or` object, and that `OR` condition is covered, then `coveredClauses` will only contain one `OR` object. In this case, `coveredClauses.size() == 1` while `conditions.numChildren() >= 1` (depending on how many children are nested within the `OR` condition. Thus, we need `coveredClauses.contains(conditions)`.

On the other hand, when `conditions` is an `And` object, then `coveredClauses` will contain the nested conditions that it cover. Under this scenario, `coveredClauses.size() == conditions.numChildren` is equivalent to `coveredClauses.containsAll(conditions)`.

This bug does not affect correctness but does harm to performance (extra in-memory filtering step is done in `QueryProcessor` even if unnecessary).

Reply:
>  Under this scenario, coveredClauses.size() == conditions.numChildren is equivalent to coveredClauses.containsAll(conditions).

Just wanted to double check this statement. Does it mean that this condition can be replaced by the next condition?
```
coveredClauses.size() == conditions.numChildren() || coveredClauses.contains(conditions)
```
Probably not but I wanted to double check because if the result is the same than `equality` operation will be faster to compute than `containsAll` operation. 
Also, `containsAll` will return `true` if `conditions.size() == 0` or if simply the amount of `conditions` elements is smaller than the amount of `coveredClauses` elements but they all equal. Thus, wanted to double check that it is the expected behavior

Reply:
Case `conditions.size() == 0` and case `conditions.numChildren() < coveredClauses.size()` are never gonna happen in theory. I will change it back to `coveredClauses.size() == conditions.numChildren()` for performance reason.